### ISSUE TITLE
Django tests in Docker container cannot be made to run in Travis (but run fine when container built locally)

### ISSUE THREAD

**Issue by MikeTheCanuck** [travis-ci/travis-ci#7473]
# Summary

I've spent a few days now troubleshooting how to get the Django test framework to run inside a Docker container via Travis CI.

Our organization (Hack Oregon) is implementing a pattern of Django-based APIs that will run in containers on AWS EC2 Container Service.  As part of the build of each Hack Oregon project's container, we wish to test that the endpoints built in Django REST Framework (DRF) are running.

At present, I have been unable (in Travis) to both (a) get the Docker container built and launched and (b) have the Django application running so that we are able to test it.

All the same code works great when run from my local Mac - I'm able to see the entire operation run from start to finish.

# Expected Result(1)
When I build the Docker container in one command, the next command I run will be able to find the container to launch it (and be able to run commands that interrogate it).

# Actual Result(1)
The Docker container that I've just built in Travis cannot be found by Docker executables to run the tests in  https://travis-ci.org/hackoregon/team-budget/builds/210649167
(see Travis build log line 479 for the error "ERROR: no container found")

(Note that this same build-container-then-access-it-in-subsequent-commands operation works fine when I build the container on my laptop.)

# Expected Result(2)
When I download a Python script to the Travis build environment and import it as a module in Django's `settings.py`, Django will be able to import the module and 

# Actual Result(2)
An ImportError for a Python module that exists in the container, but which cannot be imported at runtime by the Django app https://travis-ci.org/hackoregon/team-budget/builds/210454849
(see Travis build log line 455 for the error "ImportError: cannot import name 'project_config'")

(Note that this same download-then-import operation works fine when I build the container on my laptop.)

# Detailed Investigation
I have documented all the troubleshooting steps I've taken here:
https://github.com/hackoregon/team-budget/issues/58

I have searched the Internet high and low and been unable to find any discussion of the issues specific to this environment.  (Discussions of unrelated environments have sounded promising, but in checking all the possibilities I've been unable to defeat the issues I'm experiencing.)

**Comment by BanzaiMan** [comment]
# re 1

https://travis-ci.org/hackoregon/team-budget/builds/210649167#L462 says there is no container to start. Your script is `echo`ing `Start completed...`, but I don't think the previous command `docker-compose` actually succeeded.

# re 2

You appear to be running a Docker container from `python:3.5` https://github.com/hackoregon/team-budget/blob/71e7728f758d0e5032d3a9722bbc4f10ba1a0fa2/budget_proj/Dockerfile#L1.
We don't maintain this image. I do not know how your `docker exec` command may fail in this case.

**Comment by MikeTheCanuck** [comment]
I sincerely apologize - I was able to resolve this myself through extensive debugging of the Travis and Docker configuration (and related scripts).  Thank you for your attention to this issue and I hope it didn't cost you much time to investigate.

For anyone who happens to find this issue in the future, experiencing related problems, who is hoping for a few troubleshooting options to pursue, here's some thoughts on what was wrong and what I tried:

- I never did figure out what was the problem with issue #1 - in resolving the build issues (by making changes to the input configurations and related scripts), we never saw the issue of "no container found" reappear.  So I would consider this a symptom of root problem that, once addressed, will most likely address this symptom as well.
- when using Python to import a module that is your own custom `.py` file, if you experience an **ImportError** then ALWAYS look for a filesystem path issue.  The root cause of my own problem ultimately turned out to be that I wasn't copying the file to the folder I thought I was (it was ending up one folder higher in the hierarchy), and thus Python literally couldn't find the file to be able to import its contents.
- the dumbest troubleshooting technique might be the one that finally unearths your problem.  In my case, I inferred the path to which I was copying the file during build, but never explicitly confirmed it until very late in my efforts.  I finally inserted (into the same `.sh` script that was performing the copy of my `project_config.py` file) a series of `ls -la` commands after the `cp` command - printing out to my Travis log the specific contents of the CWD (current working directory) and all the subdirectories which I thought I knew existed in the Docker container.  This finally, unequivocally demonstrated to me where the `cp`'d file was actually landing, which made it trivial to figure out how to correct the `cp` command and finally resolve the **ImportError**.

Again, I hope this helps someone and I hope you don't spend two solid weeks trying to figure out what was wrong with Docker or Travis, when in reality the problem could be as simple as a misunderstood assumption of your code.  [Isn't that usually the case for nasty code problems?]
